### PR TITLE
Pytorch dataloader

### PR DESCRIPTION
**Pull request by LeonGuertler**
### Changes Overview

- **Preprocessing**:  
  - Moved preprocessing steps to a separate `prepare.py` script.  
  - This script handles data tokenization and storage in memory-mapped files for efficient data access.  
  - **Note**: Currently, `num_proc=1` is set during data processing to avoid multi-processing errors encountered on some machines.

- **Dataset Handling**:  
  - Introduced `datasets.py` to manage data loading with custom PyTorch Dataset classes, including:
    - `BaseDataset`
    - `BytePoolingDataset`
    - `DualBytePooling`
  - These classes support different data formats and structures, facilitating flexible data handling.

- **Sampling**:  
  - Created `samplers.py`, implementing a base random sampling method.  
  - The design allows for easy extension with additional sampling strategies.

- **Trainer Adjustments**:  
  - `BaseTrainer` now includes distinct train and validation dataloaders.  
  - Removed the train dataloader from intra-test evaluations, reducing unnecessary computation and speeding up the process.

- **Utilities**:  
  - Added a pretty print feature in `utils.py` for structured and readable output of evaluation and benchmark results.  
  - Enhanced the utility functions to support a wide range of datasets, including code datasets and competition math datasets.

- **Training Script Enhancements**:  
  - `train.py` now calls the `prepare_data` function from `prepare.py` to ensure all data is preprocessed before training begins.  
  - The script includes logic for both single GPU and distributed training (via DDP), ensuring flexibility and scalability.

This beautiful pull request summary was a curtsy of GPT-4o.

### CODE REVIEW

**Comment by LeonGuertler** [comment]
@DylanASHillier @bobbycxy

File: train.py [hunk 1]
```diff
@@ -51,6 +52,24 @@ def ddp_main(rank, world_size, cfg):
         # restore the print function
         restore_print_override(original_print)
 
+def single_gpu_main(cfg):
```
Comment (DylanASHillier) [inline train.py:55]: don't need to exist

File: configs/full_configs/baseline.yaml [hunk 2]
```diff
@@ -79,11 +69,13 @@ trainer:
...
   dataloader:
     name: standard
+  datasampling:
+    name: standard
   loss_fn:
     name: cross_entropy
 general:
   logging:
-    wandb_log: true
+    wandb_log: false
```
[the remark this answers is not in the document]
Comment (LeonGuertler) [reply configs/full_configs/baseline.yaml:78]: default should be not to populate wandb_log

File: train.py [hunk 3]
```diff
@@ -63,12 +82,21 @@ def main(cfg):
...
 
     create_folder_structure(path_config=cfg["general"]["paths"])
-    mp.spawn(
-        ddp_main,
-        args=(world_size, cfg),
-        nprocs=world_size,
-        join=True,
-    )
+
+    # process data
```
Comment (DylanASHillier) [inline train.py:86]: remove

File: trainers/build_trainers.py [hunk 4]
```diff
@@ -52,7 +56,13 @@ def ddp_setup(rank, world_size):
         weight_decay=trainer_cfg["weight_decay"],
         learning_rate=trainer_cfg["lr"],
         betas=(trainer_cfg["beta1"], trainer_cfg["beta2"]),
-    )
+    ),
```
Comment (DylanASHillier) [inline trainers/build_trainers.py:59]: ???

File: trainers/datasets.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            mode="r",
+        )
+    
+    def __getitem__(self, idx):
+        """
+        Get a batch of data from both the byte and higher token level
+        """
+        # get byte level batch
+        x_byte = torch.from_numpy((self.data_byte[idx: idx + self.context_window]).astype(np.int64))
+        #y_byte = torch.from_numpy((self.data_byte[idx + 1: idx + 1 + self.context_window]).astype(np.int64))
```
Comment (DylanASHillier) [inline trainers/datasets.py:174]: remove these comments

File: trainers/utils.py [hunk 6]
```diff
@@ -252,3 +285,37 @@ def restore_print_override(original_print):
     '''
     import builtins as __builtin__
     __builtin__.print = original_print
+
+
+
+
+# Function to print evaluation results and benchmark results
+def print_evaluation_results(iter_num, eval_results, benchmark_results):
```
Comment (DylanASHillier) [inline trainers/utils.py:293]: nice

**Review by DylanASHillier — APPROVED** [review]
LGTM but remove what i said first thx. (later we should move byte level stuff to experimental please)

**Review by LeonGuertler — COMMENTED** [review]
done

**Pull request merged**